### PR TITLE
remote address action support prefix len

### PR DESCRIPTION
**Pull request by zirain**
Signed-off-by: hejianpeng <hejianpeng2@huawei.com>

make `RemoteAddressAction` support prefix len`

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/20433
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by htuch — COMMENTED** [review]
How is this used? It seems like we're using the specific route picked to infer the CIDR range of the remote address, but that seems a pretty arbitrary association, e.g. you might have many different XFF addresses from different subnets, picking a fixed CIDR length here doesn't seem to work in general.

File: source/common/router/router_ratelimit.cc
Comment (soulxu) [inline source/common/router/router_ratelimit.cc:118]: In case of the `prefix_len` == 0, then it will get an any-address https://github.com/envoyproxy/envoy/blob/52455a260c0d1b4c6cb59aa23c290a28412794ad/source/common/network/cidr_range.cc#L154

Is this the case we expect? maybe we should limit the `prefix_len`'s value to greater than 0.

File: test/common/router/router_ratelimit_test.cc
Comment (soulxu) [inline test/common/router/router_ratelimit_test.cc:347]: Create to have unittest for prefix_len is greater than 32 for ipv4.

File: test/common/router/router_ratelimit_test.cc
Comment (soulxu) [inline test/common/router/router_ratelimit_test.cc:354]: it will great to test ipv6 case.

**Review by soulxu — CHANGES_REQUESTED** [review]

**Comment by zirain** [comment]
> How is this used? It seems like we're using the specific route picked to infer the CIDR range of the remote address, but that seems a pretty arbitrary association, e.g. you might have many different XFF addresses from different subnets, picking a fixed CIDR length here doesn't seem to work in general.

we use local rate limit in istio, doc is [here](https://istio.io/latest/docs/tasks/policy-enforcement/rate-limit/#local-rate-limit).

the address will change after pod is re-created. with this patch I can do ratelimit on cluster's pod cidr.

**Comment by htuch** [comment]
I'm still not really convinced that setting the CIDR length based on route makes sense, since the address is a property of the incoming connection and CIDR ranges will depend on source IP. I'd be interested in a second opinion here from rate limit folks @mattklein123 @rgs1 @wbpcode, as I might just be slow on the uptake.

**Comment by wbpcode** [comment]
It is reasonable requirement for me. The address can be extracted from XFF. So it would be ok to create an cidr based ratelimit entry. 

But I am not sure will this has some side effects to the rls implementation.

File: api/envoy/config/route/v3/route_components.proto
Comment (mattklein123) [inline api/envoy/config/route/v3/route_components.proto:1788]: A few points on the API:

1) What does "-1" mean in the context of an unsigned int? I'm not sure what this will evaluate to.
2) More generally, I think we have to explicitly handle v4 and v6 here. I would probably do something like have a v4_prefix_len and a v6_prefix_len, where the default if unset is /128 and /32 respectively?
3) What does "length of prefix" actually mean in practice? I assume it will just mask before sending? Can you add some examples to the API docs?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
I think this feature is reasonable but I have some API questions to get started. Thank you!

/wait

File: api/envoy/config/route/v3/route_components.proto
Comment (zirain) [reply api/envoy/config/route/v3/route_components.proto:1788]: you're right, `-1` is a bad design.
`prefix_len` comes from [here](https://github.com/envoyproxy/envoy/blob/9d5627a0879b0a029e90515137c108e1d2884bfc/api/envoy/api/v2/core/address.proto#L134), but `0` is not good as default value `In case of the prefix_len == 0, then it will get an any-address`.

`v4_prefix_len ` and `v6_prefix_len` is a good idea, I'll update soon.

Comment (mattklein123) [reply api/envoy/config/route/v3/route_components.proto:1788]: But can you clarify what this actually does? Does it mask before sending? If so can we call it something like v4_prefix_mask_len?

Comment (htuch) [reply api/envoy/config/route/v3/route_components.proto:1788]: I would like to also ask for some practical guidelines for how this would be used in practice. For example, is this only intened to be used in RFC1918 environments where subnet sizes are well understood? If it's for use on public IP, how do you pick these sizes?

Comment (zirain) [reply api/envoy/config/route/v3/route_components.proto:1788]: > But can you clarify what this actually does? Does it mask before sending? If so can we call it something like v4_prefix_mask_len?

we deploy istio with Kubernetes, envoy run as sidecar container, [here](https://istio.io/latest/docs/tasks/policy-enforcement/rate-limit/#local-rate-limit) is the doc on istio.io site.

service topology:

sidecar

right now, I can apply ratelimit rules(sleep 10/min, fortio 100/min) base on remote address(mostly podip in my case), with a custom RLS using GlobalRateLimit.
we want use this with LocalRateLimit, currently we need re-apply the `EnvoyFilter` once `fortio` or `sleep`'s podip changes.

Comment (zirain) [reply api/envoy/config/route/v3/route_components.proto:1788]: > If it's for use on public IP, how do you pick these sizes?

I may miss something, I do not think too much about the difference between public IPs and private IPs.

Comment (mattklein123) [reply api/envoy/config/route/v3/route_components.proto:1788]: I still don't understand what this does from the description. The only thing that makes sense to me is masking. Can you update the docs with specific examples? Thank you.

Comment (htuch) [reply api/envoy/config/route/v3/route_components.proto:1788]: Assuming masking as Matt points out, IMHO, the only way this works sensibly is if you have a networking environment in which traffic comes from IP addresses that are known to live in subnets of a given size. If you were operating a public Internet facing ingress gateway, it seems like this would be a pretty useless. But if this is some network you control and you know k8s pods come from well-known subnet ranges, this is just a way to ratelimit on on subnet. So, I think you need to document clearly the places this feature should be used in the Envoy docs to make this crisp.

File: api/envoy/config/route/v3/route_components.proto
Comment (zirain) [reply api/envoy/config/route/v3/route_components.proto:1791]: Agree, `MaskedRemoteAddress` would be better. I will open another PR to implement it, thanks for your help @mattklein123 .

**Comment by zirain** [comment]
@mattklein123 I've update the code, please take a review again.

File: api/envoy/config/route/v3/route_components.proto
Comment (mattklein123) [inline api/envoy/config/route/v3/route_components.proto:1798]: Update to new name, also add /32

File: api/envoy/config/route/v3/route_components.proto
Comment (mattklein123) [inline api/envoy/config/route/v3/route_components.proto:1795]:
```suggestion
      // Length of prefix mask len (e.g. 0, 32) for IPv4.
```

File: api/envoy/config/route/v3/route_components.proto
Comment (mattklein123) [inline api/envoy/config/route/v3/route_components.proto:1802]:
```suggestion
      // Length of prefix mask len for IPv6 (e.g. 0, 128), enable rate limit base on cidr format.
```

Also please add an example.

File: api/envoy/config/route/v3/route_components.proto
Comment (mattklein123) [inline api/envoy/config/route/v3/route_components.proto:1794]: This will need a release note.

File: source/common/router/router_ratelimit.cc
Comment (mattklein123) [inline source/common/router/router_ratelimit.cc:133]: This seems pretty inefficient. Can you at least add a TODO to increase the efficiency here? Seems like we can avoid a some string transform back and forth.

File: source/common/router/router_ratelimit.h
Comment (mattklein123) [inline source/common/router/router_ratelimit.h:120]: I think you can just inline this.

File: test/common/router/router_ratelimit_test.cc
Comment (mattklein123) [inline test/common/router/router_ratelimit_test.cc:364]: Add some tests where the mask length is not set.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Looks like it's on the right track, thank you!

/wait

File: api/envoy/config/route/v3/route_components.proto
Comment (zirain) [reply api/envoy/config/route/v3/route_components.proto:1794]: add release-notes need rebase of main branch due to https://github.com/envoyproxy/envoy/pull/21075 changes, I'll do that after all code are reviewed.

File: api/envoy/config/route/v3/route_components.proto
Comment (mattklein123) [inline api/envoy/config/route/v3/route_components.proto:1795]:
```suggestion
      // Length of prefix mask len for IPv4 (e.g. 0, 32).
```

File: api/envoy/config/route/v3/route_components.proto
Comment (mattklein123) [inline api/envoy/config/route/v3/route_components.proto:1802]:
```suggestion
      // Length of prefix mask len for IPv6 (e.g. 0, 128).
```

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM other than doc and release note comments.

/wait

**Comment by zirain** [comment]
@mattklein123 rebase with main and update chagelogs, thanks for your patience.

**Comment by mattklein123** [comment]
@zirain please do not force push. Just merge main and add commits. Thank you!

File: changelogs/1.23.0.yaml
Comment (mattklein123) [inline changelogs/1.23.0.yaml:83]:
```suggestion
    added support for :ref:`masked_remote_address <envoy_v3_api_field_config.route.v3.RateLimit.Action.masked_remote_address>`.
```

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with small comment, thanks.

/wait

File: api/envoy/config/route/v3/route_components.proto
Comment (kyessenov) [reply api/envoy/config/route/v3/route_components.proto:1809]: The doc seems to imply that header `x-forwarded-for` is used but the implementation just uses connection info. Is that expected?

File: changelogs/1.23.0.yaml
Comment (zirain) [reply changelogs/1.23.0.yaml]: updated.

**Comment by kyessenov** [comment]
Not blocking, but there is an opportunity to make this more general with two steps:
1. Allow any matcher input as a descriptor extension, by looking up in a secondary category.
2. Add a network or HTTP input for the masked remote address.

This way the masked address could be used in listener or route matching as well.

**Review by mattklein123 — APPROVED** [review]
LGTM. Will defer to @kyessenov for any further comments and merge. Thank you!

**Review by kyessenov — APPROVED** [review]
My comment about not using xff applies to older `remote_address` action, so it shouldn't be blocking this PR. Otherwise, LGTM.

**Pull request merged**
